plugins/discovery-ec2/src/javaRestTest/java/org/elasticsearch/discovery/ec2/DiscoveryEc2ClusterFormationTestCase.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.discovery.ec2;
+
+import org.elasticsearch.cluster.routing.Murmur3HashFunction;
+import org.elasticsearch.test.cluster.ElasticsearchCluster;
+import org.elasticsearch.test.cluster.LogType;
+import org.elasticsearch.test.rest.ESRestTestCase;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.util.HashSet;
+import java.util.regex.Pattern;
+
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.hasItem;
+
+public abstract class DiscoveryEc2ClusterFormationTestCase extends ESRestTestCase {
+
+    protected abstract ElasticsearchCluster getCluster();
+
+    @Override
+    protected String getTestRestCluster() {
+        return getCluster().getHttpAddresses();
+    }
+
+    public void testClusterFormation() throws IOException {
+        final var cluster = getCluster();
+        final var expectedAddresses = new HashSet<>(cluster.getAvailableTransportEndpoints());
+        final var addressesPattern = Pattern.compile(".* using dynamic transport addresses \\[(.*)]");
+
+        assertThat(cluster.getNumNodes(), greaterThan(1));
```

Comment:
Is it possible to make this check exact? IIUC, this is essential what the existing YAML tests do, i.e. it checks the number of nodes are expected after the cluster is up. This PR additionally checks that logs show the expected addresses which is an added-value.

Reply:
We could, but in fact it's sort of up to each test case how many nodes to start (even if today in practice they all start 2). However, if the cluster contains more than a single node then discovery must have worked, so I think this assertion should be sufficient.

Reply:
I was thinking the sub-classes can tell this method how many nodes to expect. But yeah, it's not really necessary.